File: implementations-final/augmentations/me_dissim.jl
using Pkg, Random, Distributions, StatsBase, Parameters
using Pkg
ENV["PYTHON"] = "python3.7"
Pkg.build("PyCall")

Random.seed!(123)

# structure representing a node in an isolation tree
@with_kw mutable struct It_node
	l::Union{It_node, Nothing}
	r::Union{It_node, Nothing}
	split_attr::Union{Int, Nothing}
	split_val::Union{Float64, Nothing}
	level::Int
	mass::Int = 0 
end


# !!!TESTED!!!
function get_random_itree(data_sub)

    # random_itree: define auxiliary function to implement recursion.
    function random_itree(x_in; current_height, lim)
		if current_height >= lim || size(x_in, 1) <= 1 # Base case check
        	return It_node(l=nothing, r=nothing, split_attr=nothing, split_val=nothing, level=current_height)
        else
        # Randomly select an attribute q.
        q = rand(1:size(x_in, 2))
        # Randomly select a split point p between min and max values of attribute q in X.
		min_q_x_in = minimum(x_in[:, q])
		max_q_x_in = maximum(x_in[:, q])
		if min_q_x_in == max_q_x_in
			p = min_q_x_in
		else
        	p = rand(Uniform(minimum(x_in[:, q]), maximum(x_in[:, q])))
		end
        # Get left and right subtrees.
        xl = x_in[x_in[:, q] .< p, :]
        xr = x_in[x_in[:, q] .>= p, :]
        # Recursive case
        return It_node(l=random_itree(xl, current_height=current_height+1, lim=lim),
                   r=random_itree(xr, current_height=current_height+1, lim=lim),
                   split_attr=q,
                   split_val=p,
                   level=current_height)
        end
    end

    # Build itree
    return random_itree(data_sub, current_height=0, lim=10)
end


# !!!TESTED!!!
function get_n_random_itrees(n, subs_size, data)
    random_itrees = Array{It_node, 1}(undef, n)
    # TODO: parallelize!
    for k = 1:n
        # Get a random sample of training examples to build next random itree.
        data_sub = data[sample(1:size(data, 1), subs_size, replace=false), :]
        random_itrees[k] = get_random_itree(data_sub)  # Get next random itree
    end
    return random_itrees, subs_size
end


# !!!TESTED!!!
function get_lowest_common_node_mass(itree, x1, x2)

    # If node is a leaf, return its mass.
    if itree.split_val == nothing
    	return itree.mass
    end

    # If x1 and x2 are in different subtrees, return current node's mass.
	if (x1[itree.split_attr] < itree.split_val) != (x2[itree.split_attr] < itree.split_val)
    	return itree.mass
    end

    # If both examples if left subtree, make recursive call.
	if (x1[itree.split_attr] < itree.split_val) && (x2[itree.split_attr] < itree.split_val)
    	return get_lowest_common_node_mass(itree.l, x1, x2)
    end

    # If both examples in right subtree, make recursive call.
	if (x1[itree.split_attr] >= itree.split_val) && (x2[itree.split_attr] >= itree.split_val)
    	return get_lowest_common_node_mass(itree.r, x1, x2)
    end
end


# !!!TESTED!!!
function mass_based_dissimilarity(x1, x2, itrees, subs_size)
    # In each i-tree, find lowest nodes containing both examples and accumulate masses.
    sum_masses = 0
    for i = 1:length(itrees)
    	sum_masses += get_lowest_common_node_mass(itrees[i], x1, x2)/subs_size
    end

    return (1/length(itrees)) * sum_masses  # Divide by number of space partitioning models.

end


# !!!TESTED!!!
function get_node_masses(itrees, data)

    # traverse: traverse itree with example and increment masses of visited nodes
    function traverse(example, it_node)

        # base case - in leaf
        if it_node.l == nothing && it_node.r == nothing
        	it_node.mass += 1

        # if split attribute value lower than split value
        elseif example[it_node.split_attr] < it_node.split_val
        	it_node.mass += 1
        	traverse(example, it_node.l)  # Traverse left subtree.

        # if split attribute value greater or equal to split value
        else
        	it_node.mass += 1
        	traverse(example, it_node.r)  # Traverse right subtree.
        end
    end


    # compute_masses: compute masses of nodes in itree
    function compute_masses(itree, data)
		for example_idx in 1:size(data, 1)
			traverse(data[example_idx, :], itree)
        end
    end


    # TODO: parallelize!
    for itree in itrees  # Go over itrees and set masses of nodes.
        compute_masses(itree, data)
    end
end


function get_dissim_func(num_itrees, data)
    itrees, subs_size = get_n_random_itrees(num_itrees, size(data, 1), data)
    get_node_masses(itrees, data)
	res_func = (i1, i2) -> mass_based_dissimilarity(data[i1.+1, :], data[i2.+1, :], itrees, subs_size)
	return (i1, i2, kwargs...) -> res_func.(i1, i2)
end

File: implementations-final/algorithms/augmentations/me_dissim.py
import numpy as np
from functools import partial

import pdb

class It_node:
    def __init__(self, l, r, split_attr, split_val, level, mass_comp=0):
        self.l = l                    # left subtree
        self.r = r                    # right subree
        self.split_attr = split_attr  # split attribute/dimension
        self.split_val = split_val    # split value
        self.level = level            # node level
        self.mass = 0                 # node mass (number of examples in region)

    # to_string: return a string encoding some information about the node
    def to_string(self):
        return "split_attr={0}, split_val={1}, level={2}, mass={3}".format(self.split_attr,\
                                                                 self.split_val,\
                                                                 self.level,
                                                                 self.mass)

class MeDissimilarity:

    def __init__(self, data):
        self.data = data

    def get_lowest_common_node_mass(self, itree, x1, x2):
        """ 
        Get mass of lowest node containing both x1 and x2.

        Args:
            itree               ... starting itree node
            x1                  ... example 1
            x2                  ... example 2
        Returns:
            mass of lowest node containing both x1 and x2
        """
        # 1. base case - if node is a leaf (no splitting value), return mass of node.
        if itree.split_val == None:
            return itree.mass

        # 2. base case - if split attribute value above split value for one and below split value
        # for other examples, return mass of node.
        if (x1[itree.split_attr] < itree.split_val) != (x2[itree.split_attr] < itree.split_val):
            return itree.mass

        # 1. Recursive case - both split attribute value < split value:
        # Go to left subtree.
        if x1[itree.split_attr] < itree.split_val and x2[itree.split_attr] < itree.split_val:
            return self.get_lowest_common_node_mass(itree.l, x1, x2)


        # 2. Recursive case - both split attribute value > split value:
        # Go to right subtree.
        if x1[itree.split_attr] >= itree.split_val and x2[itree.split_attr] >= itree.split_val:
            return self.get_lowest_common_node_mass(itree.r, x1, x2)


    def mass_based_dissimilarity(self, x1, x2):
        """ 
        Get mass based dissimilarity of examples x1 and x2.

        Args:
            x1                  ... example 1
            x2                  ... example 2
            random_itrees       ... list of random itrees built on training examples
            example_subset_size ... size of training example subset used to build random itrees
        Returns:
            mass based dissimilarity of examples x1 and x2
        """
        # In each i-tree, find lowest nodes containing both x and y
        # TODO: parallelize
        sum_masses = 0
        for itree in self.random_itrees:
            # Divide each sum by size of subset used to construct the trees.
            sum_masses += self.get_lowest_common_node_mass(itree, x1, x2)/self.subs_size
        return (1/len(self.random_itrees)) * sum_masses  # Divide by number of space partitioning models.


    def get_random_itree(self, data_sub):
        """ 
        Construct and return a random itree.

        Args:
            data_sub: subset of training examples on which to build the itree
        Returns:
            an instance of an In_node class representing the root of the itree
        """

        # random_itree: define auxiliary function to implement recursion.
        def random_itree(x_in, current_height, lim):
            if current_height >= lim or x_in.shape[0] <= 1: # Base case check
                return It_node(l=None, r=None, split_attr=None, split_val=None, level=current_height)
            else:
                # Randomly select an attribute q.
                q = np.random.randint(x_in.shape[1])
                # Randomly select a split point p between min and max values of attribute q in X.
                p = np.random.uniform(np.min(x_in[:, q]), np.max(x_in[:, q]))
                # Get left and right subtrees.
                xl = x_in[x_in[:, q] < p, :]
                xr = x_in[x_in[:, q] >= p, :]
                # Recursive case
                return It_node(l=random_itree(xl, current_height+1, lim),\
                               r=random_itree(xr, current_height+1, lim),\
                               split_attr=q,\
                               split_val=p,\
                               level=current_height)

        # Build itree
        return random_itree(data_sub, current_height=0, lim=10)


    def get_n_random_itrees(self, n, subs_size):
        """ 
        Construct and return n random itrees in a list.

        Args:
            n: number of random itrees to construct
            data: data on which to build the itree
            sub_size: size of subset of examples used to build the tree
        Returns:
            n random itrees in a list
        """
        random_itrees = np.empty(n, dtype=object)  # Allocate list for storing the trees.
        # TODO: parallelize!
        for k in np.arange(n):
            # Get a random sample of training examples to build next random itree.
            data_sub = self.data[np.random.choice(self.data.shape[0], subs_size, replace=False), :]
            random_itrees[k] = self.get_random_itree(data_sub)  # Get next random itree 
        self.random_itrees = random_itrees
        self.subs_size = subs_size


    def get_node_masses(self):
        """ 
        Set mass property of each node by traversing trees with each instance
        in training set.

        Args:
            data: training data
            random_itrees: list of random itrees to traverse and set mass property
        Returns:
            list of random itrees with set mass property
        """

        # traverse: traverse itree with example and increment masses of visited nodes
        def traverse(example, it_node):
            # base case - in leaf
            if it_node.l == None and it_node.r == None:
                it_node.mass += 1
            # if split attribute value lower than split value
            elif example[it_node.split_attr] < it_node.split_val:
                it_node.mass += 1
                traverse(example, it_node.l)  # Traverse left subtree.
            # if split attribute value greater or equal to split value
            else:
                it_node.mass += 1
                traverse(example, it_node.r)  # Traverse right subtree.

        # compute_masses: compute masses of nodes in itree
        def compute_masses(itree):
            for example in self.data:
                traverse(example, itree)

        # TODO: parallelize!
        for itree in self.random_itrees:  # Go over itrees and set masses of nodes.
            compute_masses(itree)

    def get_dissim_func(self, num_itrees):
        """
        get function that takes two examples and computes their mass based dissimilarity

        Args:
            data : Array[np.float64] -- training examples
            num_itrees : np.int -- number of i-trees to use
            mass_based_dissimilarity : Callable[[Array[np.float64], Array[np.float64], Array[object], np.int], np.float64] -- 
                function that implements mass based dissimilarity on learned i-trees.
        Returns:
            Callable[[Array[np.float64], Array[np.float64]], np.float64] -- learned dissimilarity function
        """
        self.get_n_random_itrees(num_itrees, self.data.shape[0])
        self.get_node_masses()

        def dissim_func(x1, x2):
            if x1.ndim == 1 and x2.ndim == 1:
                return self.mass_based_dissimilarity(x1, x2)
            elif x1.ndim == 1 and x2.ndim == 2:
                dissim_part = partial(self.mass_based_dissimilarity, x1)
                return np.apply_along_axis(dissim_part, 1, x2)
            elif x1.ndim == 2 and x2.ndim == 1:
                dissim_part = partial(self.mass_based_dissimilarity, x2)
                return np.apply_along_axis(self.mass_based_dissimilarity, 1, x1)
            elif x1.ndim == 2 and x2.ndim == 2:
                res = np.array(x1.shape[0], dtype=float)
                for idx, (r1, r2) in enumerate(zip(x1, x2)):
                    res[idx] = self.mass_based_dissimilarity(r1, r2)

        return dissim_func

# If running as main script, perform simple test.
if __name__ == "__main__":
    import numpy as np
    import scipy.io as sio
    import pdb
    examples = sio.loadmat('./data/examples.mat')['examples']
    me = MeDissimilarity(examples)
    dissim_func = me.get_dissim_func(10)
    res = dissim_func(examples[0, :], examples[0, :])
    print(res)
